File: backend/src/main/java/com/carbonfootprint/platform/mobile/dto/EnrichedTransaction.java
package com.carbonfootprint.platform.mobile.dto;

import com.carbonfootprint.platform.activity.model.ActivityCategory;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PositiveOrZero;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.HashMap;
import java.util.Map;

/**
 * Data Transfer Object representing a pre-enriched transaction sent by the Android Companion App.
 *
 * <p>This DTO contains core transaction details and optional carbon hints that
 * were extracted on-device via local ML models or rule engines.
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class EnrichedTransaction {

    @NotBlank(message = "Transaction ID cannot be blank")
    private String transactionId;

    @NotBlank(message = "Merchant cannot be blank")
    private String merchant;

    @NotNull(message = "Amount is required")
    @PositiveOrZero(message = "Amount cannot be negative")
    private BigDecimal amount;

    @NotBlank(message = "Currency cannot be blank")
    private String currency;

    @NotNull(message = "Category is required")
    private ActivityCategory category;

    private String unit;
    private String location;

    @NotNull(message = "occurredAt timestamp is required")
    private Instant occurredAt;

    private String description;
    
    /**
     * Pre-resolved hints such as activityType, transportMode, fuelType, etc.
     */
    @Builder.Default
    private Map<String, Object> carbonHints = new HashMap<>();
    
    /**
     * Additional flexible attributes.
     */
    @Builder.Default
    private Map<String, Object> metadata = new HashMap<>();
}

File: android/app/src/main/java/com/carbonwise/connect/ingestion/model/EnrichedTransaction.kt
package com.carbonwise.connect.ingestion.model

import com.carbonwise.connect.ingestion.enrichment.ConfidenceLevel

/**
 * Immutable model representing an enriched transaction.
 */
data class EnrichedTransaction(
    // Original transaction
    val merchant: String?,
    val amount: Double?,
    val currency: String?,
    val sourceApp: String?,
    val timestamp: Long,
    val rawNotification: String,
    
    // Merchant enrichment
    val merchantType: String?,
    
    // Category enrichment
    val category: String?,
    
    // Carbon enrichment
    val carbonHint: String?,
    val transportMode: String?,
    val fuelType: String?,
    val energyType: String?,
    val purchaseType: String?,
    
    // Confidence
    val overallConfidence: Double,
    val confidenceLevel: ConfidenceLevel,
    val confidenceBreakdown: Map<String, Double>,
    
    // Metadata
    val enrichmentVersion: String,
    val processedAt: Long,
    val processingTimeMs: Long,
    val stageMetricsMs: Map<String, Long>,
    val warnings: List<String>,
    val reasoning: List<String>
)
